### ISSUE TITLE
Nested ternaries are just misunderstood.

### ISSUE THREAD

**Issue by Qix-** [sindresorhus/xo#12]
So I just made a comment here about consistent-this hindering actual functionality - something code styling should do conservatively. Here is something else pretty relevant.

Ternary operations can be performed safely and visibly if nested. As well, there now becomes an issue of branching.

Take the following code:

```javascript
result.file = opts.dash ? '-' : platform === 'linux' ?
	'/dev/stdin' : '/proc/self/fd/0';
```

or even this

```javascript
result.file = opts.dash
	? '-'
	: platform === 'linux'
		? '/proc/self/fd/0'
		: '/dev/stdin';

```

But this results in the following error:

```
index.js
  27:17  error  Do not nest ternary expressions  no-nested-ternary
```

Seriously? Here is the alternative.

```javascript
var inputPath = platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0';
result.file = opts.dash ? '-' : inputPath;
```

Which is *not* correct logic (`inputPath` should only be evaluated if `opts.dash` is not truthy).

The other alternative is

```javascript
var inputPath;
if (opts.dash) {
	inputPath = '-';
} else {
	inputPath = platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0';
}
```

Which, come on, seriously? >.> So unnecessary. A nested ternary operation was just fine and readable. I understand this is something I can disable in the package.json but I want to be as, uh... "orthodox" as possible. Plus, turning it off will not (should not) interfere with any existing code.

**Comment by Qix-** [comment]
If anything, what are the thoughts on why this shouldn't be allowed?

**Comment by stevemao** [comment]
I never find ternary operation readable. Nested ternary operation is probably worse :(
Also JavaScript operator precedence could be pretty tricky.

**Comment by markogresak** [comment]
I must agree with @stevemao. For me, even a simple ternary is potential red flag, because they often execute differently than I first anticipated.

For example (I know it's stupid, but serves to show my usual usage):

``` js
var someBool = true, anotherBool = false, val = 5;
result = someBool ? getVal1() : anotherBool ? getVal2() : getVal3() + val;
```

In this case, I wanted to express: if `someBool` then execute `getVal1()` else if `anotherBool` call `getVal2()` and otherwise use `getVal3()` and add `val` to received result.
I couldn't come up with an example where the problem I often faced isn't visible immediately like here, but I hope you get the idea. The problem is that without adding the brackets around actual ternary (only `get*`), the `val` will be added only to the last one, if it's executed.
But consider a more complex ternary. Things get messy quickly and when someone actually puts this in their code, it's really hard to know later and determining if it's a bug or a deliberate thing can be difficult. 

As a basic rule to great code, write code which one can just scan through and quickly comprehend what is going on. You make code reviews go faster or just make it easier for someone else checking out your project. And also, it's way simpler for you to spot errors. The whole purpose of using strict linters is to easily debuggable code. As for your ternary example, I had to stop for a moment and think what you're doing. I stupidly reassured myself about this many times, then learned it was bad the hard way, which was spending whole day getting familiar with my own code. And of course not getting paid to do that.

**tl;dr**: nested ternaries can get nasty fast. I say keep no-nested-ternary rule as it is. Write understandable code, not voodoo magic crap you will forget about later on.

**Comment by Qix-** [comment]
I agree with what you're saying, but I still find it a stretch.

```javascript
anotherBool ? getVal2() : getVal3() + val;
```

To me, this is immediately clear you're adding (or concatenating) `getVal3() + val` as its own expression; this is a simple property of ternaries that carry over from all other languages. It's a control structure in and of itself. I find it hard to believe someone would misinterpret that.

I can scan through and look at the code in the OP just fine. It's immediately clear, thanks to formatting (which passes just fine in most styling checks I've used) and reduces branching a bit while also reducing LOC (i.e. no if statement with mandatory `{}`'s).

They can get nasty, and that's where a complexity constraint should be at play, definitely. A one-level-deep nesting, as in the OP, shouldn't be ostracized for being too complex.

The only thing I can think about being against ternary operators *as a whole* is that you can't debug them as easily (no breakpoints).

**Comment by markogresak** [comment]
@Qix- Damn it, I knew exactly this would happen. I emphasised that it's a simple and stupid example, asking to try consider something more complex going on there. And yet it seems you didn't even try. 
It's obvious that you can't miss it in a short line, but believe me, people are capable of making true abominations with something as simple as nested ternary. Then someone has to spend few hours debugging it. 

There might not be a big problem in your example either, it's just the fact that someone else reading your code might not be as familiar. Reading the whole context of code would probably be understandable faster, but that's not the point. Allowing nested ternaries means it's inevitable that someone will start writing unreadable expression. The point of linters is to prevent us from writing any potentially bad code.

Also, here's how I would write your original code:

``` js
var inputPath = '-';
if (!opts.dash) {
  inputPath = platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0';
}
```

You get rid of nested ternary and don't have to make a huge if-else.

**Comment by Qix-** [comment]
No but now you have an extra assignment ;) Also, ad-homs deserve no place on GitHub. Lighten up a bit!

**Comment by sindresorhus** [comment]
I'm torn on this as I sometimes use nested ternaries, but gonna go with the strict version as it's beneficial to more people. In the end, we should optimize for readability.

I always quoted my sub-ternaries though, for clarity:

```js
result.file = opts.dash ? '-' : (platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0');
```

-

This is still up for discussion if anyone else has any additional arguments.

**Comment by Qix-** [comment]
I agree that parenthesizing them is an improvement. I usually just format them in the way I did in the OP if they're simple enough.

<sup>And for the record, the piece of code has since been refactored ;)</sup>

**Comment by markogresak** [comment]
@Qix- What are you talking about? Cheer up, no one is attacking you, especially not personally. :smile: 
I'm also not sure about the second part... What else can i do about your code to make you feel I put some thought and effort into it? 

I also wanted to point out what @sindresorhus said, using parentheses makes things clearer. But I changed my mind to point out I prefer writing clear code right away, rather than thinking if what I just wrote is clear enough or should I clarify it more.

**Comment by Qix-** [comment]
@markogresak @sindresorhus it appears nested ternaries **with parenthesis** is still disallowed ;)

**Comment by sindresorhus** [comment]
@Qix- Yes, in my comment, I meant I *used to use* parens around nested ternaries. Maybe worth opening an ESLint issue about having an excepting option in the rule for when parens are used? https://github.com/eslint/eslint/issues

**Comment by Qix-** [comment]
Ah I see what you mean. Will do :+1:.
